tiledb/sm/filesystem/mem_filesystem.h

```diff
@@ -212,7 +234,7 @@ class MemFilesystem {
   /* ********************************* */
 
   /* The node that represents the root of the directory tree. */
-  tdb_unique_ptr<FSNode> root_;
+  shared_ptr<FSNode> root_;
```

Comment:
`shared_ptr` has inferior performance and unless we really need shared ownership we should avoid using them. Would the solution I suggested for avoiding the 2 step construction with `init` remove the need for changing form `unique_ptr` to `shared_ptr` as well?

Reply:
I think it's best to stick with `shared_ptr` here. Because `MemFilesystem` isn't default constructed, adding in a move constructor on a `unique_ptr<FSNode>&&` doesn't seem viable. We will have to use move semantics throughout `MemFilesystem`, `FSNode`, `VFS` and the C API to properly retrieve the root and use it to initialize a new memfs. For now, I'll push the code that's using `shared_ptr` to directly initialize but am happy to discuss further!

Reply:
Thanks for the changes in the constructor, looks good!
Regarding the `shared_ptr` vs `unique_ptr` I have to admit I am deeply confused on the desired ownership here. Once this is clear we can decide on how we can implement it.
@ihnorton @eric-hughes-tiledb Can you please share your thoughts on this issue? To recap:
`memfs` is currently designed to be tightly coupled to a **context**, or more accurately to a **VFS**. To create an array we pass in a `Context`, which has a `storage_manager()` instance which has a `VFS()` instance. `VFS()` has `MemoryFileSystem` as a member, which is initialized upon `VFS` creation and is used to "store" this non-persistent array that is actually a tree.
So far we have been using `unique_ptr` throughout the `MemoryFileSystem` tree as the `VFS` instance of this context was the only owner of this object. As a result an array written using ctx1 wouldn't be available/visible using ctx2, so a using a global context is suggested when using memfs.
However, @bekadavis9 found and fixed a bug, where although read and write work fine, we weren't able to `remove` or `ls` on this filesystem. And this is were it gets confusing to me: 
The reason is that to `remove` or `ls` an array we typically create a new VFS instance and execute the command there, e.g. from the example Beka added:
```
void remove_array(std::string array_name) {
  // Initialize a VFS object
  VFS vfs(ctx_);

  // Ensure the memfs directory has been initialized
  std::vector<std::string> uris = vfs.ls(array_name);
  for (auto uri : uris) {
    std::cout << uri << std::endl;

  // Clean up
  vfs.remove_dir(array_name);
```
Creating a new `VFS` using the same context "breaks" this assumption of single ownership of `memfs`, as we need to be able to access the filesystem through any number of `VFS`s that might be created, so we need to introduce shared ownership with the known performance penalties it comes with.

I am a bit confused with all these, please let me know what's your opinion on this and if you have suggestions on how we should proceed and what's the right ownership model.

Reply:
I'll clarify a bit here on the ownership of `MemFS` and the need for a global `Context`. 
In essence, we have a single, global `Context` and are passing _reference_ to the `MemFS` from an old `VFS` to the new, in-test `VFS`. 

> To create an array we pass in a Context, which has a storage_manager() instance which has a VFS() instance. VFS() has MemoryFileSystem as a member

This is correct. _However_ there are 2 different "ownership" issues at hand:

1. The `MemFS` is  created on a single `Context` and cannot be written to / read from on a different one. This was the original bug and is a previously-unstated invariant of the `MemFilesystem` class. 

2. `ctx_->storage_manager()->vfs() != VFS vfs(ctx_)`. Presuming we have a single, global context, creating a VFS in-test _default initializes_ the memfs, losing the array data. As such, we cannot call `ls`, `remove_dir`, etc. because the memfs is empty. Note that other filesystems (posix, s3, etc) are default initialized but ultimately have `init` called on them to initialize member variables from passed-in arguments. This is what we are trying to accomplish by grabbing the root from the "original" vfs and _passing_ it to the new vfs. The memfs still exists where it should on the `Context` for array manipulation but we can now see it on the `VFS` for work-tree manipulation as well. 

Hopefully that clears some things up but @ihnorton please expand as needed

I should add that if we so decide, the VFS "stuff" may not be entirely necessary...? As I mention above, it's really just adding an in-test copy of the memfs for user visibility. It's helpful to make available `ls` but I believe the `remove_dir` API will not be invoked on the "original" memfs. I could certainly be wrong so _anyone_ feel free to correct me.

Reply:
>The reason is that to remove or ls an array we typically create a new VFS instance 

This feels very broken to me. What I'm seeing is that there's a discussion about how to complicate code in one place to work around some existing code that's badly implemented in another place. My basic opinion on this is very succinct: Don't do that.

It looks from where I'm sitting that the right solution looks like this:
1. Declare that each memory-filesystem is associated with a single `VFS` instance. To the extent that each context only uses a single VFS, you've also got the one-per-context rule as well.
2. To the extent that a context is using multiple VFS, that context is using the VFS badly. Declare that code to be defective and change it.
    * It should already be considered an error to open up a new VFS just to do an operation. It's bad practice to reinitialize unnecessarily.
3. Part of the problem is that the `VFS` constructor is publicly visible. It probably shouldn't be. If the whole idea is that the VFS should be obtained through a context (or a storage manager), then it should be obtained _only_ that way.
    * The simple expedient to change is to declare a few friends for `class VFS` and make all its constructors private. I did a quick test compile just making the default constructor private and it only breaks a few things. It looks like you may only need three: one for the API (which allows direct use), one for `StorageManager`, and one for the heap memory tracer.
    * If there are tests that open a new VFS directly, they'd need to be changed to get their VFS from a context.
4. Eliminate `VFS::init()` and move its code to a constructor. It's only called twice outside of tests.
   * Once is in C API function `tiledb_vfs_alloc`. Here it's called soon after constructing it. No issue rewriting that code.
   * The other is in `StorageManager::init`. Here it's called _on the very next line_ after constructing it. Even less issue rewriting it. (Eliminate `StorageManager::init` in a separate pass.)
   * 9 occurrences in tests. Rewrite those.

Going this direction avoids all the complication of sharing memory-filesystem. I'd wager that the current memfs code can't handle multiprocessing correctly anyway.

If, down the road, there's a need to share memfs state for some reason, we can talk about how to do that.

Reply:
Added comment on SC Story to reflect updated work item after Eric's comment & discussion with Seth